python/ray/experimental/sgd/model.py

```diff
@@ -7,16 +7,61 @@ class Model(object):
     """Your class must implement this interface to be used with Ray SGD.
 
     This supports any form of input pipeline: it is up to you to define it
-    using TensorFlow. The only requirements are that the loss and optimizer
-    attributes must be defined.
-
+    using TensorFlow.
     For an example implementation, see tfbench/test_model.py
-
-    Attributes:
-        loss (tf.Tensor): Loss function to minimize.
-        optimizer (tf.train.Optimizer): Optimizer to use to minimize the loss.
     """
 
+    def build(self):
+        """Build customized tensorflow model
+
+        Users should setup variables such as loss or in get_loss function
+        """
+        raise NotImplementedError('build of %s is not implemented' % self.__class__.__name__)
+
+    def get_loss(self):
+        """Return loss of the model
+
+        Returns:
+            loss
+        """
+        raise NotImplementedError('get_loss of %s is not implemented' % self.__class__.__name__)
+
+    #TODO support complex way of updating gradient, e.g. using different optimizers
+    def get_optimizer(self):
+        """Return optimizer for the model
+
+        Returns:
+            optimizer
+        """
+        raise NotImplementedError('get_optimizer of %s is not implemented' \
+                % self.__class__.__name__)
+
+    def get_metrics(self):
+        """Return metrics of the model
+
+        Returns:
+            metrics(dict): e.g. {'accuray': accuracy_tensor}
+        """
+        return {}
+
+    def get_predict_score(self):
+        """Return tensor for predict score
+
+        Returns:
+            score: tensor for predicted probability
+        """
+        raise NotImplementedError('get_predict_score of %s is not implemented' \
+                % self.__class__.__name__)
+
+    def get_predict_label(self):
```

Comment:
How are these different from metrics? Should they be just one method?

Reply:
These two interfaces are left for future usage (Maybe I should delete them currently and I will). Certain metrics such f1, AUC cannot be calculated using tensors. The interfaces are added so we can retrieve them in SGDWorker. `get_metrics` is used to get tensors related to the metrics of the training batch, such as loss, or any metric that user cares during training.